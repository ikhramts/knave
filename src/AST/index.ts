export class TokenLocation {
    start: Position;
    end: Position;
}

export class Position {
    offset: number;
    line: number;
    column: number;
}

export class FileDeclarations {
    declarations: Declaration[]

    constructor(declarations: Declaration[]) {
        this.declarations = declarations;
    }
}

export type Declaration = Table

export class Table {
    header : TableHeader;
    body: TableBody

    constructor(header: TableHeader, body: TableBody) {
        this.header = header;
        this.body = body;
    }
}

export class TableHeader {
    name: string

    constructor(name: string) {
        this.name = name;
    }
}

export class TableBody {
    lines: TableBodyLine[]

    constructor(lines: TableBodyLine[]) {
        this.lines = lines;
    }
}

export type TableBodyLine = ColumnDefinition;

export class ColumnDefinition {
    name: string;
    type: string;
    constraints: ColumnConstraints;

    constructor(name: string, type: string, constraintList: ColumnConstraints[]) {
        this.name = name;
        this.type = type;
        
        let constraints = new ColumnConstraints({});
        for (let constraint of constraintList) {
            (Object as any).assign(constraints, constraint);
        }

        this.constraints = constraints;
    }
}

export class ColumnConstraints {
    primaryKey?: boolean;
    notNull?: boolean;

    constructor(init: Partial<ColumnConstraints>) {
        (<any>Object).assign(this, init);
    }
}