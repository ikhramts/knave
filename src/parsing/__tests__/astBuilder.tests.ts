import { parseFile } from '../astBuilder';
import * as AST from '../../AST';

describe("parseFile()", () => {
    it("Should procude a FileDeclarations object", () => {
        expect.assertions(2);
        return parseFile(inputFile('basicCreateTable.sql')).then(fileDeclarations => {
            expect(fileDeclarations).not.toBeNull();
            expect(fileDeclarations.declarations).not.toBeNull();
        })
    })

    it("Should read table name", () => {
        expect.assertions(1);
        return parseFile(inputFile('basicCreateTable.sql')).then(fileDeclarations => {
            let table = fileDeclarations.declarations[0];
            expect(table.header.name).toBe("Products");;
        })
    })

    it("Should read all table lines", () => {
        expect.assertions(1);
        return parseFile(inputFile('basicCreateTable.sql')).then(fileDeclarations => {
            let table = fileDeclarations.declarations[0];
            expect(table.body.lines.length).toBe(2);
        })
    });

    it("Should read column name", () => {
        expect.assertions(1);
        return parseFile(inputFile('basicCreateTable.sql')).then(fileDeclarations => {
            let column = fileDeclarations.declarations[0].body.lines[0];
            expect(column.name).toBe("id");
        });
    });

    it("Should parse ATS correctly", () => {
        expect.assertions(1);
        return parseFile(inputFile('basicCreateTable.sql')).then(fileDeclarations => {
            expect(fileDeclarations).toMatchSnapshot();
        });
    });
});

function inputFile(pathFromTests:string) : string {
    return "src/parsing/__tests__/inputs/" + pathFromTests;
}