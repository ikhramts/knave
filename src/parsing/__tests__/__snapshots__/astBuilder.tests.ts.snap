// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFile() Should parse ATS correctly 1`] = `
FileDeclarations {
  "declarations": Array [
    Table {
      "body": TableBody {
        "lines": Array [
          ColumnDefinition {
            "constraints": ColumnConstraints {
              "notNull": true,
              "primaryKey": true,
            },
            "name": "id",
            "type": "bigint",
          },
          ColumnDefinition {
            "constraints": ColumnConstraints {
              "notNull": true,
            },
            "name": "lineNumber",
            "type": "smallint",
          },
        ],
      },
      "header": TableHeader {
        "name": "Products",
      },
    },
  ],
}
`;
